### PR TITLE
[CF-450] Refactor discovery/model.py

### PR DESCRIPTION
**Pull request by antonf**
This patch decompose old discovery model into two separate entitis. One
is model which is responsible for model serialization, deserialization,
validation and storing this data into database. Other part is only
responsible for discovery. I hope this simplifies the code, make data
conversion from representation returned by OpenStack clients to
our internal representation explicit (no more magic) and makes discovery
process extensible.

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.os.estimation.procedures
cloudferry/lib/os/estimation/procedures.py:41: [E1101(no-member), estimate_copy] Module 
'cloudferry.lib.os.discovery.nova' has no 'Server' member
cloudferry/lib/os/estimation/procedures.py:57: [E1101(no-member), estimate_copy] Module 
'cloudferry.lib.os.discovery.cinder' has no 'Volume' member
cloudferry/lib/os/estimation/procedures.py:63: [E1101(no-member), estimate_copy] Module 
'cloudferry.lib.os.discovery.glance' has no 'Image' member
cloudferry/lib/os/estimation/procedures.py:99: [E1101(no-member), show_largest_servers] 
Module 'cloudferry.lib.os.discovery.nova' has no 'Server' member
cloudferry/lib/os/estimation/procedures.py:116: [E1101(no-member), 
show_largest_unused_resources] Module 'cloudferry.lib.os.discovery.nova' has no 
'Server' member
cloudferry/lib/os/estimation/procedures.py:126: [E1101(no-member), 
show_largest_unused_resources] Module 'cloudferry.lib.os.discovery.cinder' has no 
'Volume' member
cloudferry/lib/os/estimation/procedures.py:142: [E1101(no-member), 
show_largest_unused_resources] Module 'cloudferry.lib.os.discovery.glance' has no 
'Image' member
************* Module cloudferry.lib.os.migrate.keystone
cloudferry/lib/os/migrate/keystone.py:33: [E0602(undefined-variable), 
CreateTenant.migrate] Undefined variable 'keystone'
************* Module cloudferry.lib.os.discovery.nova
cloudferry/lib/os/discovery/nova.py:215: [W0621(redefined-outer-name), 
_need_image_membership] Redefining name 'image' from outer scope (line 22)
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:66: [W0621(redefined-outer-name), 
ReserveImage.migrate] Redefining name 'image' from outer scope (line 19)
cloudferry/lib/os/migrate/glance.py:75: [E0602(undefined-variable), 
ReserveImage.migrate] Undefined variable 'glance'
cloudferry/lib/os/migrate/glance.py:242: [E1101(no-member), MigrateImageMember.migrate] 
Class 'ImageMember' has no 'make' member
************* Module cloudferry.model.compute
cloudferry/model/compute.py:29: [E1101(no-member), Flavor.equals] Instance of 'Flavor' 
has no 'object_id' member
************* Module cloudferry.model.identity
cloudferry/model/identity.py:28: [E1101(no-member), Tenant.equals] Instance of 'Tenant' 
has no 'name' member
************* Module cloudferry.model.storage
cloudferry/model/storage.py:30: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'device' member
************* Module tests.model.test_model
tests/model/test_model.py:27: [W0621(redefined-outer-name), 
ExampleReferenced.create_object] Redefining name 'uuid' from outer scope (line 15)
tests/model/test_model.py:131: [W0621(redefined-outer-name), ModelTestCase._make_id] 
Redefining name 'uuid' from outer scope (line 15)

**Comment by MirantisWorkloadMobility** [comment]
Merge check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3389/

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3397/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.lib.os.discovery.nova
cloudferry/lib/os/discovery/nova.py:215: [W0621(redefined-outer-name), 
_need_image_membership] Redefining name 'image' from outer scope (line 22)
************* Module cloudferry.lib.os.migrate.keystone
cloudferry/lib/os/migrate/keystone.py:33: [E0602(undefined-variable), 
CreateTenant.migrate] Undefined variable 'keystone'
************* Module cloudferry.model.identity
cloudferry/model/identity.py:28: [E1101(no-member), Tenant.equals] Instance of 'Tenant' 
has no 'name' member
************* Module cloudferry.model.compute
cloudferry/model/compute.py:29: [E1101(no-member), Flavor.equals] Instance of 'Flavor' 
has no 'object_id' member
************* Module cloudferry.model.storage
cloudferry/model/storage.py:30: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'device' member
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:66: [W0621(redefined-outer-name), 
ReserveImage.migrate] Redefining name 'image' from outer scope (line 19)
cloudferry/lib/os/migrate/glance.py:75: [E0602(undefined-variable), 
ReserveImage.migrate] Undefined variable 'glance'
cloudferry/lib/os/migrate/glance.py:242: [E1101(no-member), MigrateImageMember.migrate] 
Class 'ImageMember' has no 'make' member
************* Module tests.model.test_model
tests/model/test_model.py:27: [W0621(redefined-outer-name), 
ExampleReferenced.create_object] Redefining name 'uuid' from outer scope (line 15)
tests/model/test_model.py:131: [W0621(redefined-outer-name), ModelTestCase._make_id] 
Redefining name 'uuid' from outer scope (line 15)

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3402/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.discover
cloudferry/discover.py:77: [W0107(unnecessary-pass), Discoverer.find_obj] Unnecessary 
pass statement
************* Module cloudferry.lib.os.discovery.nova
cloudferry/lib/os/discovery/nova.py:215: [W0621(redefined-outer-name), 
_need_image_membership] Redefining name 'image' from outer scope (line 22)
************* Module cloudferry.lib.os.migrate.keystone
cloudferry/lib/os/migrate/keystone.py:33: [E0602(undefined-variable), 
CreateTenant.migrate] Undefined variable 'keystone'
************* Module cloudferry.model.identity
cloudferry/model/identity.py:28: [E1101(no-member), Tenant.equals] Instance of 'Tenant' 
has no 'name' member
************* Module cloudferry.model.compute
cloudferry/model/compute.py:29: [E1101(no-member), Flavor.equals] Instance of 'Flavor' 
has no 'object_id' member
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:66: [W0621(redefined-outer-name), 
ReserveImage.migrate] Redefining name 'image' from outer scope (line 19)
cloudferry/lib/os/migrate/glance.py:75: [E0602(undefined-variable), 
ReserveImage.migrate] Undefined variable 'glance'
cloudferry/lib/os/migrate/glance.py:242: [E1101(no-member), MigrateImageMember.migrate] 
Class 'ImageMember' has no 'make' member
************* Module cloudferry.model.storage
cloudferry/model/storage.py:30: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'device' member
************* Module tests.model.test_model
tests/model/test_model.py:27: [W0621(redefined-outer-name), 
ExampleReferenced.create_object] Redefining name 'uuid' from outer scope (line 15)
tests/model/test_model.py:131: [W0621(redefined-outer-name), ModelTestCase._make_id] 
Redefining name 'uuid' from outer scope (line 15)

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.discover
cloudferry/discover.py:77: [W0107(unnecessary-pass), Discoverer.find_obj] Unnecessary 
pass statement
************* Module cloudferry.lib.os.discovery.nova
cloudferry/lib/os/discovery/nova.py:215: [W0621(redefined-outer-name), 
_need_image_membership] Redefining name 'image' from outer scope (line 22)
************* Module cloudferry.lib.os.migrate.keystone
cloudferry/lib/os/migrate/keystone.py:33: [E0602(undefined-variable), 
CreateTenant.migrate] Undefined variable 'keystone'
************* Module cloudferry.model.identity
cloudferry/model/identity.py:28: [E1101(no-member), Tenant.equals] Instance of 'Tenant' 
has no 'name' member
************* Module cloudferry.model.compute
cloudferry/model/compute.py:29: [E1101(no-member), Flavor.equals] Instance of 'Flavor' 
has no 'object_id' member
************* Module cloudferry.model.storage
cloudferry/model/storage.py:30: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'device' member
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:66: [W0621(redefined-outer-name), 
ReserveImage.migrate] Redefining name 'image' from outer scope (line 19)
cloudferry/lib/os/migrate/glance.py:75: [E0602(undefined-variable), 
ReserveImage.migrate] Undefined variable 'glance'
cloudferry/lib/os/migrate/glance.py:242: [E1101(no-member), MigrateImageMember.migrate] 
Class 'ImageMember' has no 'make' member
************* Module tests.model.test_model
tests/model/test_model.py:27: [W0621(redefined-outer-name), 
ExampleReferenced.create_object] Redefining name 'uuid' from outer scope (line 15)
tests/model/test_model.py:131: [W0621(redefined-outer-name), ModelTestCase._make_id] 
Redefining name 'uuid' from outer scope (line 15)

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3403/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.model.compute
cloudferry/model/compute.py:29: [E1101(no-member), Flavor.equals] Instance of 'Flavor' 
has no 'object_id' member
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:243: [E1101(no-member), MigrateImageMember.migrate] 
Class 'ImageMember' has no 'make' member
************* Module cloudferry.model.identity
cloudferry/model/identity.py:28: [E1101(no-member), Tenant.equals] Instance of 'Tenant' 
has no 'name' member
************* Module cloudferry.model.storage
cloudferry/model/storage.py:30: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'device' member
************* Module tests.model.test_model
tests/model/test_model.py:27: [W0621(redefined-outer-name), 
ExampleReferenced.create_object] Redefining name 'uuid' from outer scope (line 15)
tests/model/test_model.py:131: [W0621(redefined-outer-name), ModelTestCase._make_id] 
Redefining name 'uuid' from outer scope (line 15)

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/3410/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.model.compute
cloudferry/model/compute.py:29: [E1101(no-member), Flavor.equals] Instance of 'Flavor' 
has no 'object_id' member
************* Module cloudferry.model.identity
cloudferry/model/identity.py:28: [E1101(no-member), Tenant.equals] Instance of 'Tenant' 
has no 'name' member
************* Module cloudferry.lib.os.migrate.glance
cloudferry/lib/os/migrate/glance.py:243: [E1101(no-member), MigrateImageMember.migrate] 
Class 'ImageMember' has no 'make' member
************* Module cloudferry.model.storage
cloudferry/model/storage.py:30: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'device' member
************* Module tests.model.test_model
tests/model/test_model.py:27: [W0621(redefined-outer-name), 
ExampleReferenced.create_object] Redefining name 'uuid' from outer scope (line 15)
tests/model/test_model.py:131: [W0621(redefined-outer-name), ModelTestCase._make_id] 
Redefining name 'uuid' from outer scope (line 15)

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferry.model.storage
cloudferry/model/storage.py:30: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'server' member
cloudferry/model/storage.py:32: [E1101(no-member), Attachment.equals] Instance of 
'Attachment' has no 'device' member
************* Module tests.model.test_model
tests/model/test_model.py:27: [W0621(redefined-outer-name), 
ExampleReferenced.create_object] Redefining name 'uuid' from outer scope (line 15)
tests/model/test_model.py:131: [W0621(redefined-outer-name), ModelTestCase._make_id] 
Redefining name 'uuid' from outer scope (line 15)

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferry/model/__init__.py [hunk 1]
```diff
@@ -15,26 +15,9 @@
 Model module contain tools to define cloud object schema and a way to store
```
Comment (SVilgelm) [inline cloudferry/model/__init__.py]: i think using something like `base.py` module instead of putting the code into `__init__.py` is better.

Comment (antonf) [inline cloudferry/model/__init__.py]: Can you explain why? I think that it's better to have several {{__init__.py}} is better then several base modules. Reason: probability of error is less when you don't have modules with same names in different packages.

**Pull request closed without merging**